eth/consensus/clique/snapshot_manager.py

```diff
@@ -0,0 +1,311 @@
+                    current_header.block_number, current_header.parent_hash)
+            except SnapshotNotFound as e:
+                current_header = self._lookup_header(current_header.parent_hash)
+
+                if is_checkpoint(current_header.block_number, self._epoch_length):
+                    new_snapshot = self._create_snapshot_from_checkpoint_header(current_header)
+                    break
+                else:
+                    parents.append(current_header)
+            else:
+                break
+
+        for parent in reversed(parents):
+            new_snapshot = self.apply(new_snapshot, parent)
+
+        new_snapshot = self.apply(new_snapshot, header)
+
+        if is_checkpoint(header.block_number, self._epoch_length):
+            self.logger.debug2("Persisting checkpoint snapshot at %s", header.block_number)
+            self.persist_snapshot(new_snapshot)
+
+        return new_snapshot
+
+    def get_snapshot(self, block_number: int, block_hash: Hash32) -> Snapshot:
+        """
+        Return a ``Snapshot`` if it exists in memory, on-disk or can be computed directly from
+        a header that serves as a checkpoint. Otherwise raise a ``SnapshotNotFound`` error.
+        """
+        # We first try to find the snapshot in memory
+        if block_hash in self._snapshots:
+            return self._snapshots[block_hash]
+
+        if is_checkpoint(block_number, self._epoch_length):
+            try:
+                # We might have it saved on disk
+                return self.get_snapshot_from_db(block_hash)
+            except SnapshotNotFound:
+                pass
+
+            try:
+                # Otherwise, we can retrieve it on the fly
+                header = self._chain_db.get_block_header_by_hash(block_hash)
+            except HeaderNotFound:
+                raise SnapshotNotFound(f"Can not get snapshot for {block_hash} at {block_number}")
+            else:
+                if header.block_number != block_number:
+                    raise SnapshotNotFound(
+                        f"Can not get snapshot for {block_hash} at {block_number}"
+                    )
+                else:
+                    return self._create_snapshot_from_checkpoint_header(header)
+
+        raise SnapshotNotFound(f"Can not get snapshot for {block_hash} at {block_number}")
+
+    def add_snapshot(self, mutable_snapshot: MutableSnapshot) -> Snapshot:
+        """
+        Retrieve a ``Snapshot`` from the given ``mutable_snapshot`` and add it to the local cache.
+        """
+        snapshot = mutable_snapshot.get_frozen_clone()
+        self._snapshots[snapshot.block_hash] = snapshot
+        return snapshot
+
+    def persist_snapshot(self, snapshot: Snapshot) -> None:
+        """
+        Persist the given snapshot to the database.
+        """
+        key = make_snapshot_lookup_key(snapshot.block_hash)
+        with self._chain_db.db.atomic_batch() as db:
```

Comment:
I don't think we need the `atomic_batch` here since this only sets a single value (which is an atomic action)